# Genome Coverage Visualizer for Metagenomics

## Online Demo:
For more information please visit http://mohammedalser.bilkent.edu.tr/MiCoP.html !
![alt text](https://github.com/smangul1/miCoP/blob/master/Genome_Coverage_Visualizer/Interactive_Plot/MiCoP_Demo.png)

## Usage:
```
/u/project/zarlab/malser/MiCoP/Scripts/MiCoP_GenomeCoverageVisualizer.sh /u/scratch2/scratch2/m/malser/MergedEuPathDB [full directory to your read set] [full directory of Scripts folder] [full directory to output, even if not exists] [window size] [Homology Information Folder generated by homology_per_genome.py] [Kmer Size for generating short segments of RefSeq database] [Unique coverage percentage threshold, to ignore any genome that has less coverage then the threshold]
```

## Running a test:
```
qrsh -l i,h_rt=24:00:00,h_data=25G
screen
module load python/3.6.1
module load bwa
module load samtools
sed -i -e 's/\r$//' /u/project/zarlab/malser/MiCoP/Scripts/MiCoP_GenomeCoverageVisualizer.sh
chmod +x /u/project/zarlab/malser/MiCoP/Scripts/MiCoP_GenomeCoverageVisualizer.sh
/u/project/zarlab/malser/MiCoP/Scripts/MiCoP_GenomeCoverageVisualizer.sh /u/scratch2/scratch2/m/malser/MergedEuPathDB /u/home/galaxy/collaboratory/serghei/MetaSUB-Inter-City-Challenge/data/SRR3546361.fastq /u/project/zarlab/malser/MiCoP/Scripts /u/project/zarlab/malser/MiCoP 1 /u/scratch2/scratch2/m/malser/HomologyInformation/ 30 30
```
